enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,349 @@
+`machineconfiguration.openshift.io` API group.
+
+The new custom resource definition is described in detail in
+https://github.com/openshift/api/pull/1609.
+
+### Implementation Details/Notes/Constraints
+
+Starting with version 4.14 of OpenShift CRI-O will have the capability to pin
+certain images (see [this](https://github.com/cri-o/cri-o/pull/6862) pull
+request for details). That capability will be used to pin all the images
+required for the upgrade, so that they aren't garbage collected by kubelet and
+CRI-O.
+
+In addition when the CRI-O service is upgraded and restarted it removes all the
+images. This used to be done by the `crio-wipe` service, but is now done
+internally by CRI-O. It can be avoided setting the `version_file_persist`
+configuration parameter to "", but that would affect all images, not just the
+pinned ones. This behavior needs to be changed in CRI-O so that pinned images
+aren't removed, regardless of the value of `version_file_persist`.
+
+The changes to pin the images will be done in a file inside the
+`/etc/crio/crio.conf.d` directory. To avoid potential conflicts with files
+manually created by the administrator the name of this file will be the name of
+the `PinnedImageSet` custom resource concatenated with the UUID assiged by the
+API server. For example, if the custom resource is this:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-pinned-images
+  uuid: 550a1d88-2976-4447-9fc7-b65e457a7f42
+spec:
+  pinnedImages:
+  - quay.io/openshift-release-dev/ocp-release@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  ...
+```
+
+Then the complete path will be this:
+
+```txt
+/etc/crio/crio.conf.d/my-pinned-images-550a1d88-2976-4447-9fc7-b65e457a7f42.conf
+```
+
+The content of the file will the `pinned_images` parameter containing the list
+of images:
+
+```toml
+pinned_images=[
+  "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  ...
+]
+```
+
+In addition to the list of images to be pinned, the `PinnedImageSet` custom
+resource will also contain a node selector. This is intended to support
+different sets of images for different kinds of nodes. For example, to pin
+different images for control plane and worker nodes the user could create two
+`PinnedImageSet` custom resources:
+
+```yaml
+# For control plane nodes:
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-control-plane-pinned-images
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/control-plane: ""
+  pinnedImages:
+  ...
+
+---
+
+# For worker nodes:
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-control-plane-pinned-images
```

Comment:
```suggestion
  name: my-worker-pinned-images
```
You preface this by saying they're different but then use the same PinnedImageSet name

Reply:
Thanks for pointing that out, it is fixed now.